src/lib/chawathe98_matcher.rs

```diff
@@ -46,35 +46,2262 @@
+    pub fn new(base: Arena<T, FromNodeId>, diff: Arena<T, ToNodeId>) -> MappingStore2<T> {
+        MappingStore2 {
+            from_arena: RefCell::new(base),
+            to_arena: RefCell::new(diff),
+            list_edges: RefCell::new(Vec::new()),
+            list_node_inserted: RefCell::new(Vec::new()),
+            list_node_deleted: RefCell::new(Vec::new()),
+            all_edge_cost: RefCell::new(CostEdge::new(1, 1, 1, 1, 1, 1, 1, 1)),
+        }
+    }
+    /// Implementation of the update of the cost edge.
+    /// This will renew the cost value for each edge type.
+    pub fn update_cost(
+        &mut self,
+        cost_update: usize,
+        cost_insert: usize,
+        cost_delete: usize,
+        cost_move: usize,
+        cost_copy: usize,
+        cost_glue: usize,
+        cost_null: usize,
+        cost_ok: usize,
+    ) {
+        self.all_edge_cost = RefCell::new(CostEdge::new(
+            cost_update,
+            cost_insert,
+            cost_delete,
+            cost_move,
+            cost_copy,
+            cost_glue,
+            cost_null,
+            cost_ok),
+        );
+    }
+
+    /// Push New Edge
+    pub fn push(
+        &mut self,
+        new_from_node: NodeId<FromNodeId>,
+        new_to_node: NodeId<ToNodeId>,
+        new_value: usize,
+        new_edge_type: EdgeType,
+    ) {
+        let new_edge = Edge::new(new_from_node, new_to_node, new_value, new_edge_type);
+        self.list_edges.borrow_mut().push(new_edge);
+    }
+    /// Remove Edge
+    pub fn remove(&mut self, index: usize) {
+        self.list_edges.borrow_mut().remove(index);
+    }
+    /// Check contains the from_node
+    pub fn contains_from(&self, from: NodeId<FromNodeId>) -> bool {
+        let list_edges = self.list_edges.borrow();
+        for edge in list_edges.iter() {
+            if edge.from_node == from {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /// Check if the edges contains
+    ///     1) from node id &
+    ///     2) To node id
+    ///     3) Edge Type
+    pub fn contains_edge(
+        &self,
+        from: usize,
+        to: usize,
+        edge_type: EdgeType,
+        list_edges: &mut Vec<Edge>,
+    ) -> bool {
+        for edge in list_edges {
+            if edge.from_node.id() == from && edge.to_node.id() == to && edge.edge_type == edge_type
+                {
+                    return true;
+                }
+        }
+        return false;
+    }
+
+    /// Check if the edge contains the to_node
+    pub fn contains_to(&self, to: NodeId<ToNodeId>) -> bool {
+        let list_edges = self.list_edges.borrow();
+        for edge in list_edges.iter() {
+            if edge.to_node == to {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /// Print for all the pre order traversal tree2: <Vec<NodeId<ToNodeId>>>
+    pub fn print_preorder_traversal(
```

Comment:
There are a few places where you have different "print" type functions. I think for now let's leave them, but in general in Rust we would implement the `fmt::Display` or `fmt::Debug` trait on a struct, and then it would be possible to just say `println!("{:?}", my_struct_instance);`. It might be worth raising an issue for this and coming back to it another day, but I think this PR is big enough that we don't want to make this particular change here.